### PR TITLE
Partitioning support for feature group and dataset

### PR DESCRIPTION
**Pull request by yogesh266**

### CODE REVIEW

File: ads/feature_store/feature_group.py [hunk 1]
```diff
@@ -325,6 +327,37 @@ def with_primary_keys(self, primary_keys: List[str]) -> "FeatureGroup":
...
+        Returns
+        -------
+        FeatureGroup
+            The FeatureGroup instance (self)
+        """
+        return self.set_spec(
+            self.CONST_PARTITION_KEYS,
+            {
+                self.CONST_ITEMS: [
+                    {self.CONST_PRIMARY_KEY_NAME: partition_key}
```
Comment (KshitizLohia) [inline ads/feature_store/feature_group.py:355]: can we change this to `CONST_NAME`

File: ads/feature_store/execution_strategy/delta_lake/delta_lake_service.py [hunk 2]
```diff
@@ -216,9 +226,9 @@ def save_delta_dataframe(
 
             dataframe.write.format("delta").options(
                 **self.get_delta_write_config(feature_delta_write_option_config)
-            ).mode(dataframe_ingestion_mode).saveAsTable(table_name)
+            ).mode(dataframe_ingestion_mode).partitionBy(delta_partition_keys).saveAsTable(table_name)
```
Comment (KshitizLohia) [inline ads/feature_store/execution_strategy/delta_lake/delta_lake_service.py:229]: what occurs when saveAsTable is used with an empty list? also, we should focus on implementing integration test cases for partition keys.

File: ads/feature_store/entity.py [hunk 3]
```diff
@@ -314,24 +316,27 @@ def _build_feature_group(
...
-        expectation_suite: ExpectationSuite = None,
-        expectation_type: ExpectationType = ExpectationType.NO_EXPECTATION,
-        statistics_config: Union[StatisticsConfig, bool] = True,
-        transformation_id: str = None,
-        name: str = None,
-        description: str = None,
-        compartment_id: str = None,
+            self,
+            primary_keys: List[str],
+            partition_keys: List[str] = None,
```
Comment (KshitizLohia) [inline ads/feature_store/entity.py:321]: add a unit test/integration test for when None is provided vs empty list is provided.

File: ads/feature_store/execution_strategy/delta_lake/delta_lake_service.py [hunk 4]
```diff
@@ -216,11 +235,15 @@ def save_delta_dataframe(
...
                 **self.get_delta_write_config(feature_delta_write_option_config)
-            ).mode(dataframe_ingestion_mode).saveAsTable(table_name)
-        else:
-            dataframe.write.format("delta").mode(dataframe_ingestion_mode).saveAsTable(
+            ).mode(dataframe_ingestion_mode).partitionBy(
+                delta_partition_keys
+            ).saveAsTable(
                 table_name
             )
+        else:
```
Comment (KshitizLohia) [inline ads/feature_store/execution_strategy/delta_lake/delta_lake_service.py:243]: we might need to check how cases for `__execute_delta_merge_insert_update_all` and `__execute_delta_merge_insert_update` should be handled with partition keys

File: ads/feature_store/feature_group.py (hunk 1 above)
Comment (yogesh266) [reply ads/feature_store/feature_group.py]: Resolved

File: ads/feature_store/execution_strategy/delta_lake/delta_lake_service.py (hunk 2 above)
Comment (yogesh266) [reply ads/feature_store/execution_strategy/delta_lake/delta_lake_service.py]: it will not partition the data if we pass empty list

**Comment by KshitizLohia** [comment]
The initial columns [A, B] were evolved into the final columns [A, B, C], partitioned by A. The following are the tested modes:

1. Upsert
2. Overwrite
3. Append

**Review by KshitizLohia — DISMISSED** [review]
please run integration tests and paste in the pr before merging the pr.

**Review by KshitizLohia — APPROVED** [review]
